distinctBase type trait for distinct types

cooldome opened the pull request:
In quest for migration to ARC I was looking for `mashal` module replacement. I have tried a number of serialization libraries in last 2 days and all of them had one thing in common. They can't handle distinct types while `marshal` had no issues with such types. Major reason  Nim doesn't provide much help to work with distinct types. Most libraries use `fieldPairs` iterator with operator overloading and `when` statements as a fallback.

`when T is distinct` already works for distinct types the last missing part of the puzzle is to get base type of distinct.  I have added baseType() type trait to help with that. I have tried it on the samson library and it worked well.

Also, I found `isNamedTuple` trait useful for serialization, but because the way it was implemented it can't be used in `when` statements which is problem, so I decided to reimplement in `evalTypeTrait` .

Araq commented:
Needs a changelog entry and a ``.since`` annotation or the corresponding ``when`` statement.

dumjyl commented:
How is this different than `sugar.distinctBase`?
This one doesn't desymbolize the result, which I have had issues with before. Anything else aside from being magic?

Also the signature + name is rather generic for only working with distinct types.
Something like sugar's `distinctBase(T: typedesc)` or `baseType(T: typedesc[distinct])` and let overload resolution provide the error.

cooldome commented:
@dumjyl. Good comment thank you. I wasn't aware of `sugar.distinctBase`,
What I will do I will harmonize it, I  will move distinctBase to typetraits because it is where it belongs and I will use name distinctBase then we don't need since annotation.

timotheecour commented on tests/metatype/ttypetraits.nim:
```diff
@@ -58,3 +58,29 @@ var xvar: selectType(string)
+  doAssert(b2 is distinct)
+  doAssert(c2 is distinct)
+
+  doAssert($distinctBase(typeof(AA))) == "seq[int]")
```
* why `typeof(AA)` when AA is already a type ?
* no need for stringification, instead: doAssert distinctBase(AA) == seq[int]

cooldome replied:
do like to test operation chaining here and I am using `$`. It suits well.
Tests like `distinctBase(AA) is seq[int]` are already done in tsugar.nim, no need to repeat them. 
`tsugar` was ported in this PR to use `distinctBase` from `typetraits` module. 
Though I will change test to use `typeof(variable)` instead, this one wasn't tested

timotheecour replied:
> do like to test operation chaining here and I am using $. It suits well.

then maybe do it for 1 test, not all tests; stringification of types is being tested in dedicated tests already;
note that your test is strictly speaking weaker since you could have `$A == $B` without `A is B`

> Tests like distinctBase(AA) is seq[int] are already done in tsugar.nim, no need to repeat them.

I don't follow the logic; currently both typetraits.distinctBase is being tested in tsugar.nim (which has `import sugar except distinctBase` and `import ttypetraits`) and in `ttypetraits`. 
typetraits.distinctBase should be tested in exactly one place (ttypetraits)

Also, are there any semantic differences between `typetraits.distinctBase` and `sugar.distinctBase`? if yes, we should use a different name; if no, `sugar.distinctBase` should be made a deprecated alias for `typetraits.distinctBase`  instead of keeping its implementation, and its test in tsugar.nim should be moved to ttypetraits.nim

cooldome replied:
Bla bla bla. 

In the end `$distinctBase(c) == "xxx"` caught bugs that `distinctBase(c) is XX` couldn't. 
Tested.

timotheecour commented:
+1 also for turning `isNamedTuple` into a magic; IIRC it was causing some issue and is definitely more efficient as a magic
(I wrote both the original isNamedTuple and distinctBase as macros but implementation is indeed simpler / more efficient as a magic)

cooldome commented:
I just discovered that `ttypetraits.nim` is a disabled test. I need to move mytests elsewhere

narimiran commented on changelog.md:
```diff
@@ -17,6 +17,8 @@
   backends, and it is compatible with Python's behavior,
   e.g. `formatFloat(3.14159, precision = 0)` is now `3`, not `3.`.
 - Global variable `lc` has been removed from sugar.nim.
+- `distincBase` has been moved from sugar.nim to typetraits and now implemented as
```
Minor, typo: `distincBase` should be `distinctBase` (added `t`).

(on the following line too)

timotheecour commented:
> I just discovered that ttypetraits.nim is a disabled test. I need to move my tests elsewhere

there's already a ttypetraits2.nim, you can use that

dumjyl commented:
@timotheecour 
> /cc @dumjyl what do you mean? can you please explain by showing how they differ with an example ?

`sugar.distinctBase` replaces `nnkSym` with `nnkIdent` so the wrong type may be bound. 

Untested, but its essential from:
```nim
# mod_a
type Bar* = string
type Foo* = distinct Bar

# mod_b
import mod_a, sugar
type Bar = int
type Detail = distincBase(Foo)
assert Detail is string
```

I believe some have had problems with constructors of `nnkSym`, but `nnkSym` -> `nnkIdent` is a one way transformation that can be done by the user.

cooldome commented:
PR is ready. All comments address.
I have consolidated all `distinctBase` tests into `ttypetraits`. `tsugar` ended up being almost empty, so I moved all remaining tests and removed `tsugar`.

timotheecour commented:
@dumjyl 
> sugar.distinctBase replaces nnkSym with nnkIdent so the wrong type may be bound.

thanks! indeed, I checked, the assert fails, it returns int instead of string, so sugar.distincBase has a pre-existing bug (gets fooled in that case by lookup with same names)

## sugar.distinctBase is recursive, unlike typetraits.distinctBase
I had argued in https://github.com/nim-lang/Nim/pull/8531#discussion_r207764878 that `distinctBase` should be non-recursive (with an option for recursive), however @Araq disagreed and wanted only recursive;
however, in this PR, typetraits.distinctBase is actually non-recursive:
```nim
    type T = distinct int
    doAssert distinctBase(T) is int
    doAssert: not compiles(distinctBase(int))
    type T2 = distinct T
    doAssert distinctBase(T2) is int # fails for `typetraits.distinctBase`
```

so we should arrive at a new consensus; 

in any case, IMO we shouldn't keep old implementation around (especially because it has a bug, and especially if it has same name but different semantics)

## proposal
* make sugar.distinctBase a deprecated alias of typetraits.distinctBase:
* and re-introduce the `recursive` flag in typetraits.distinctBase
```nim
# typetraits.nim
proc distinctBase*(T: typedesc, recursive: static[bool] = false): typedesc {.magic: "TypeTrait".}

# sugar.nim
import std/typetraits
template distinctBase*(T): untyped {.deprecated: "use typetraits.distinctBase".} = sugar.distinctBase(T, recursive = true) # for compatibility
```

cooldome commented:
Tests are green, ready for review

timotheecour commented:
one more thing to note (that doesn't require any change in this PR):

* before PR: this would fail:  Error: attempt to access a nil address
after PR: this works
```nim
  import std/typetraits
  proc fun2[T](a: openArray[T]): bool =
    isNamedTuple (int, T, T)
  static:
    doAssert not fun2(@[1,2])
```

* i had come up with another implementation that doesn't fail with above testcase and didn't use `compiles` but I still prefer this PR's version because it doesn't add a dependency on typetraits (and is likely easier on the compiler)

```nim
import std/macros

macro isNamedTuple*(T: typedesc): bool =
  # note: getTypeInst would work, but not getType
  var n = T
  n = getTypeImpl(n)
  while true:
    case n.kind
    of nnkBracketExpr: n=getTypeImpl(n[1])
    of nnkTupleTy: return newLit true
    of nnkTupleConstr: return newLit false
    of nnkSym:
      doAssert false # or maybe `return newLit false`
    else:
      return newLit false
```
